File: solutions/typescript/linked-list/1/linked-list.ts
class ListNode<T> {
  value: T;
  next: ListNode<T> | null = null;
  prev: ListNode<T> | null = null;

  constructor(value: T) {
    this.value = value;
  }
}

export class LinkedList<T> {
  private head: ListNode<T> | null = null;
  private tail: ListNode<T> | null = null;
  private length: number = 0;

  public push(element: T): void {
    const newNode = new ListNode(element);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail!.next = newNode;
      newNode.prev = this.tail;
      this.tail = newNode;
    }
    this.length++;
  }

  public pop(): T | undefined {
    if (!this.tail) return undefined;
    const value = this.tail.value;
    if (this.head === this.tail) {
      this.head = null;
      this.tail = null;
    } else {
      this.tail = this.tail.prev;
      this.tail!.next = null;
    }
    this.length--;
    return value;
  }

  public shift(): T | undefined {
    if (!this.head) return undefined;
    const value = this.head.value;
    if (this.head === this.tail) {
      this.head = null;
      this.tail = null;
    } else {
      this.head = this.head.next;
      this.head!.prev = null;
    }
    this.length--;
    return value;
  }

  public unshift(element: T): void {
    const newNode = new ListNode(element);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      newNode.next = this.head;
      this.head.prev = newNode;
      this.head = newNode;
    }
    this.length++;
  }

  public delete(element: T): void {
    let current = this.head;
    while (current) {
      if (current.value === element) {
        if (current === this.head) {
          this.shift();
        } else if (current === this.tail) {
          this.pop();
        } else {
          current.prev!.next = current.next;
          current.next!.prev = current.prev;
          this.length--;
        }
        return;
      }
      current = current.next;
    }
  }


  public count(): number {
    return this.length;
  }
}

File: solutions/kotlin/linked-list/1/src/main/kotlin/LinkedList.kt
class Deque<T> {
    private class Node<T>(val value: T) {
        var next: Node<T>? = null
        var prev: Node<T>? = null
    }

    private var head: Node<T>? = null
    private var tail: Node<T>? = null

    fun push(value: T) {
        val newNode = Node(value)
        if (head == null) {
            head = newNode
            tail = newNode
        } else {
            tail?.next = newNode
            newNode.prev = tail
            tail = newNode
        }
    }

    fun pop(): T? {
        if (tail == null) return null
        val value = tail?.value
        if (head == tail) {
            head = null
            tail = null
        } else {
            tail = tail?.prev
            tail?.next = null
        }
        return value
    }

    fun unshift(value: T) {
        val newNode = Node(value)
        if (head == null) {
            head = newNode
            tail = newNode
        } else {
            newNode.next = head
            head?.prev = newNode
            head = newNode
        }
    }

    fun shift(): T? {
        if (head == null) return null
        val value = head?.value
        if (head == tail) {
            head = null
            tail = null
        } else {
            head = head?.next
            head?.prev = null
        }
        return value
    }
}
